src/Microsoft.AspNetCore.Http.Connections/Internal/Transports/HttpUpgradeTransport.cs

```diff
@@ -0,0 +1,285 @@
+
+                    var flushResult = await _application.Output.FlushAsync();
+
+                    // We canceled in the middle of applying back pressure
+                    // or if the consumer is done
+                    if (flushResult.IsCanceled || flushResult.IsCompleted)
+                    {
+                        break;
+                    }
+                }
+            }
+            catch (ConnectionResetException ex)
+            {
+                // Client has closed the WebSocket connection without completing the close handshake
+                Log.ClosedPrematurely(_logger, ex);
+            }
+            catch (OperationCanceledException)
+            {
+                // Ignore aborts, don't treat them like transport errors
+            }
+            catch (Exception ex)
+            {
+                if (!_aborted)
+                {
+                    _application.Output.Complete(ex);
+
+                    // We re-throw here so we can communicate that there was an error when sending
+                    // the close frame
+                    throw;
+                }
+            }
+            finally
+            {
+                // We're done writing
+                _application.Output.Complete();
+            }
+        }
+
+        private async Task StartSending(Stream stream)
+        {
+            Exception error = null;
+
+            try
+            {
+                while (true)
+                {
+                    var result = await _application.Input.ReadAsync();
+                    var buffer = result.Buffer;
+
+                    // Get a frame from the application
+
+                    try
+                    {
+                        if (result.IsCanceled)
+                        {
+                            break;
+                        }
+
+                        if (!buffer.IsEmpty)
+                        {
+                            try
+                            {
+                                Log.SendPayload(_logger, buffer.Length);
+
+                                await stream.WriteAsync(buffer);
+                            }
+                            catch (Exception ex)
+                            {
+                                if (!_aborted)
+                                {
+                                    Log.ErrorWritingFrame(_logger, ex);
+                                }
+                                break;
+                            }
+                        }
+                        else if (result.IsCompleted)
+                        {
+                            break;
+                        }
+                    }
+                    finally
+                    {
+                        _application.Input.AdvanceTo(buffer.End);
+                    }
+                }
+            }
+            catch (WebSocketException ex) when (ex.WebSocketErrorCode == WebSocketError.ConnectionClosedPrematurely)
```

Comment:
When would this ever happen? You're writing to a stream now, not a WebSocket.

Reply:
You could argue this should be handled in the application layer and not in the transport.